client/src/components/TagField.js

```diff
@@ -51,205 +67,156 @@ class TagField extends Component {
+   * @param {string} key
+   * @returns {boolean}
    */
-  fetchOptions(input) {
-    const { optionUrl, labelKey, valueKey } = this.props;
-    const fetchURL = url.parse(optionUrl, true);
-    fetchURL.query.term = input;
-
-    return fetch(url.format(fetchURL), { credentials: 'same-origin' })
-      .then((response) => response.json())
-      .then((json) => json.items.map(
-        (item) => ({
-          [labelKey]: item.Title,
-          [valueKey]: item.Value,
-          Selected: item.Selected,
-        })
-      ));
-  }
+  const valueInOptions = (checkValue, checkOptions, key) => {
+    // eslint-disable-next-line no-restricted-syntax
+    for (const item of checkOptions) {
+      if (checkValue === item[key]) {
+        return true;
+      }
+    }
+    return false;
+  };
 
   /**
    * Check if a new option can be created based on a given input
    * @param {string} inputValue
-   * @param {array|object} value
+   * @param {array|object} checkValue
    * @param {array} currentOptions
    * @returns {boolean}
    */
-  isValidNewOption(inputValue, value, currentOptions) {
-    const { valueKey } = this.props;
-
+  const isValidNewOption = (inputValue, checkValue, currentOptions) => {
     // Don't allow empty options
     if (!inputValue) {
       return false;
     }
 
     // Don't repeat the currently selected option
-    if (Array.isArray(value)) {
-      if (this.valueInOptions(inputValue, value, valueKey)) {
+    if (Array.isArray(checkValue)) {
+      if (valueInOptions(inputValue, checkValue, valueKey)) {
         return false;
       }
-    } else if (inputValue === value[valueKey]) {
+    } else if (inputValue === checkValue[valueKey]) {
       return false;
     }
 
     // Don't repeat any existing option
-    return !this.valueInOptions(inputValue, currentOptions, valueKey);
+    return !valueInOptions(inputValue, currentOptions, valueKey);
+  };
+
+  const optionAttributes = lazyLoad
+    ? { loadOptions: getOptions }
+    : { options };
+
+  let DynamicSelect = SelectComponent;
+  if (lazyLoad && creatable) {
+    DynamicSelect = AsyncCreatableSelectComponent;
+  } else if (lazyLoad) {
+    DynamicSelect = AsyncSelectComponent;
+  } else if (creatable) {
+    DynamicSelect = CreatableSelectComponent;
   }
 
-  /**
-   * Check if a value is in an array of options already
-   * @param {string} value
-   * @param {array} options
-   * @param {string} valueKey
-   * @returns {boolean}
-   */
-  valueInOptions(value, options, valueKey) {
-    // eslint-disable-next-line no-restricted-syntax
-    for (const item of options) {
-      if (value === item[valueKey]) {
-        return true;
-      }
-    }
-    return false;
+  // Update the value to passthrough with the kept state provided this component is not
+  // "controlled"
+  const passThroughAttributes = { ...passThroughAttributesFromProps };
+  if (!isControlled()) {
+    passThroughAttributes.value = stateValue;
+    passThroughAttributes.optionUrl = optionUrl;
```

Comment:
I think `optionUrl` should be added whether it's controlled or not (used to be included implicitly due to not being desconstructed out of the `props` variable)

Reply:
Opps, fixed